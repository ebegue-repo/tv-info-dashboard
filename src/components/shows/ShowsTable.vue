<template>
    <div class="divide-y divide-gray-200 overflow-hidden rounded-lg bg-white shadow">
        <search-bar @searchEvent="searchEventHandler" class="mt-7"/>
        <div class="px-4 py-5 sm:p-6">
            <div class="flow-root">
                <div class="-my-2 -mx-6 overflow-x-auto lg:-mx-8">
                    <div class="inline-block min-w-full py-2 align-middle sm:px-6 lg:px-8">
                    <table v-if="shows.length > 0" class="min-w-full divide-y divide-gray-300">
                        <thead>
                        <tr>
                            <th v-for="item in headers" scope="col" class="px-3 py-3.5 text-left text-sm font-semibold text-gray-900">{{ item }}</th>
                        </tr>
                        </thead>
                        <tbody class="divide-y divide-gray-200 bg-white cursor-pointer">
                        <tr v-for="show in shows.slice(0,6)" :key="show.id" @click="clickOnRow(show.arrayIndex)">
                            <td class="whitespace-nowrap px-3 py-2 text-sm text-gray-500">
                            <img v-if="show.image" class="h-15 w-10 rounded" :src="show.image.original" alt="" />
                            <img v-else class="h-15 w-10 rounded" src="/default.jpg" alt="" />
                            </td>
                            <td class="whitespace-nowrap px-3 py-2 text-sm text-gray-500">
                            <div class="text-gray-500">{{ show.name }}</div>
                            </td>
                            <td class="whitespace-nowrap px-3 py-2 text-sm text-gray-500">
                            <div class="text-gray-500">{{ showGenres(show.genres) }}</div>
                            </td>
                            <td class="whitespace-nowrap px-3 py-2 text-sm text-gray-500">
                            <span class="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800">{{ show.rating.average }}</span>
                            </td>
                        </tr>
                        </tbody>
                    </table>
                    <h1 v-else class="font-bold tracking-tight sm:text-xl lg:text-2xl text-center">No results for this search</h1>
                  </div>
                </div>
            </div>
        </div>
    </div>
</template>

<script lang="ts">
    import SearchBar from '../generic/SearchBar.vue';

    export default {
      name: 'GenericTable',
      components: {
        SearchBar
      },
      props: {
        headers: {
          type: Array,
          required: true
        },
        shows: {
          type: Array,
          required: true
        }
      },
      methods: {
        showGenres: function(genres: String[]) {
          let finalString = "";

          genres.forEach(item => {
            finalString += item + ", ";
          });

          return finalString.slice(0, -2);
        },

        clickOnRow: function(arrayIndex: number) {
          this.$emit('clickOnRowEvent', arrayIndex);
        },

        searchEventHandler: function(searchInput: string) {
          this.$emit('searchEvent', searchInput);
        }
      }
    }
</script>