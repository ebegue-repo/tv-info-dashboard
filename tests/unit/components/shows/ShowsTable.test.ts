import { mount } from '@vue/test-utils';
import ShowsTable from '@/components/shows/ShowsTable.vue';
import { describe, expect, test, vi } from 'vitest';

describe('ShowsTable', () => {
    test('It should render the shows table with the proper headers and show name. Also it has to match the snapshot', () => {

        const showsArrayMock = [{
            arrayIndex: 0,
            id: 1657,
            url: "https://www.thetvdb.com/banners/show/1657",
            name: "Under the dome",
            type: "Scripted",
            language: "English",
            genres: ["Drama, Science-Fiction, Crime"],
            status: "Ended",
            runtime: 60,
            averageRuntime: 60,
            premiered: "2013-06-24",
            ended: "2015-09-10",
            officialSite: "http://www.cbs.com/shows/under-the-dome/",
            schedule: {
              "time": "22:00",
              "days": [
                  "Thursday"
              ]
            },
            rating: 6.8,
            weight: 98,
            network: {
              "id": 2,
              "name": "CBS",
              "country": {
                  "name": "United States",
                  "code": "US",
                  "timezone": "America/New_York"
              },
              "officialSite": "https://www.cbs.com/"
            },
            webChannel: null,
            dvdCountry: null,
            externals: {
              "tvrage": 25988,
              "thetvdb": 264492,
              "imdb": "tt1553656"
            },
            image: {
              "medium": "https://static.tvmaze.com/uploads/images/medium_portrait/81/202627.jpg",
              "original": "https://static.tvmaze.com/uploads/images/original_untouched/81/202627.jpg"
            },
            summary: "<p><b>Under the Dome</b> is the story of a small town that is suddenly and inexplicably sealed off from the rest of the world by an enormous transparent dome. The town's inhabitants must deal with surviving the post-apocalyptic conditions while searching for answers about the dome, where it came from and if and when it will go away.</p>",
            updated: 1631010933,
            _links: {
              "self": {
                  "href": "https://api.tvmaze.com/shows/1"
              },
              "previousepisode": {
                  "href": "https://api.tvmaze.com/episodes/185054"
              }
            }
          }]
""
        const headersMock = ["Image", "Name", "Genre", "Rating"];  

        const wrapper = mount(ShowsTable, {
            props: {
              shows: showsArrayMock,
              headers: headersMock
            }
        });

        expect(wrapper.text()).toMatch("Image");
        expect(wrapper.text()).toMatch("Name");
        expect(wrapper.text()).toMatch("Genre");
        expect(wrapper.text()).toMatch("Rating");
        expect(wrapper.text()).toMatch("Under the dome");
        expect(wrapper.html()).toMatchSnapshot();
    });
});