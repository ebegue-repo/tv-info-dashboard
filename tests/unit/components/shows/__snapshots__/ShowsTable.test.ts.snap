// Vitest Snapshot v1

exports[`ShowsTable > It should render the shows table with the proper headers and show name. Also it has to match the snapshot 1`] = `
"<div class=\\"divide-y divide-gray-200 overflow-hidden rounded-lg bg-white shadow\\">
  <div class=\\"flex justify-center mt-7\\">
    <div class=\\"mb-3 xl:w-96\\">
      <div class=\\"relative mb-4 flex w-full flex-wrap items-stretch\\"><input type=\\"search\\" class=\\"relative m-0 -mr-px block w-[1%] min-w-0 flex-auto rounded-l border border-solid border-neutral-300 bg-transparent bg-clip-padding px-3 py-1.5 text-base font-normal text-neutral-700 outline-none transition duration-300 ease-in-out focus:border-primary focus:text-neutral-700 focus:shadow-te-primary focus:outline-none dark:text-neutral-200 dark:placeholder:text-neutral-200\\" placeholder=\\"Search\\" aria-label=\\"Search\\" aria-describedby=\\"button-addon1\\"><button type=\\"submit\\" class=\\"bg-tvm-green text-white inline-flex justify-center rounded-md py-2 px-5 mr-3 text-sm font-medium shadow-sm focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2 ml-5\\">Search</button></div>
    </div>
  </div>
  <div class=\\"px-4 py-5 sm:p-6\\">
    <div class=\\"flow-root\\">
      <div class=\\"-my-2 -mx-6 overflow-x-auto lg:-mx-8\\">
        <div class=\\"inline-block min-w-full py-2 align-middle sm:px-6 lg:px-8\\">
          <table class=\\"min-w-full divide-y divide-gray-300\\">
            <thead>
              <tr>
                <th scope=\\"col\\" class=\\"px-3 py-3.5 text-left text-sm font-semibold text-gray-900\\">Image</th>
                <th scope=\\"col\\" class=\\"px-3 py-3.5 text-left text-sm font-semibold text-gray-900\\">Name</th>
                <th scope=\\"col\\" class=\\"px-3 py-3.5 text-left text-sm font-semibold text-gray-900\\">Genre</th>
                <th scope=\\"col\\" class=\\"px-3 py-3.5 text-left text-sm font-semibold text-gray-900\\">Rating</th>
              </tr>
            </thead>
            <tbody class=\\"divide-y divide-gray-200 bg-white cursor-pointer\\">
              <tr>
                <td class=\\"whitespace-nowrap px-3 py-2 text-sm text-gray-500\\"><img class=\\"h-15 w-10 rounded\\" src=\\"https://static.tvmaze.com/uploads/images/original_untouched/81/202627.jpg\\" alt=\\"\\"></td>
                <td class=\\"whitespace-nowrap px-3 py-2 text-sm text-gray-500\\">
                  <div class=\\"text-gray-500\\">Under the dome</div>
                </td>
                <td class=\\"whitespace-nowrap px-3 py-2 text-sm text-gray-500\\">
                  <div class=\\"text-gray-500\\">Drama, Science-Fiction, Crime</div>
                </td>
                <td class=\\"whitespace-nowrap px-3 py-2 text-sm text-gray-500\\"><span class=\\"inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800\\"></span></td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
